### PR TITLE
[SPEC] Add ND input data support for DepthToSpace and SpaceToDepth

### PR DESCRIPTION
**Pull request by mbencer**

### CODE REVIEW

File: src/ngraph/op/fused/depth_to_space.cpp [hunk 1]
```diff
@@ -46,59 +47,100 @@ op::DepthToSpace::DepthToSpace(const Output<Node>& data,
...
-    NGRAPH_CHECK((c % (m_blocksize * m_blocksize) == 0 && m_blocksize > 0),
-                 "SpaceToDepth: The depth axis size must be a multiple of ",
-                 "squared block_size attribute value.");
+    const size_t n_dim = data_shape.at(0);
+    const size_t c_dim = data_shape.at(1);
+    const size_t spatial_dims = data_shape.size() - 2;
+    const auto c_dim_divider = static_cast<int>(std::pow(m_blocksize, spatial_dims));
+
+    NGRAPH_CHECK(m_blocksize > 0 && c_dim % c_dim_divider == 0,
+                 "The color axis size: ",
```
Comment (postrational) [inline src/ngraph/op/fused/depth_to_space.cpp]:
```suggestion
                 "The channels axis size: ",
```

File: src/ngraph/op/fused/depth_to_space.cpp [hunk 2]
```diff
@@ -46,59 +47,100 @@ op::DepthToSpace::DepthToSpace(const Output<Node>& data,
...
+                 c_dim,
+                 " must be a multiple of ",
+                 "block_size^spatial_dims: ",
+                 c_dim_divider,
+                 " attribute value");
```
Comment (postrational) [inline src/ngraph/op/fused/depth_to_space.cpp]: I think this should be reworder to read: 
`DepthToSpace: The input data's 'channels' axis size: 20, must be a equivalent to 'block_size'^'spatial_dims': 40.`

File: src/ngraph/op/fused/depth_to_space.cpp [hunk 3]
```diff
@@ -46,59 +47,100 @@ op::DepthToSpace::DepthToSpace(const Output<Node>& data,
...
     // First we have to disperse the data from depth channel, then rearrange them
     // so as appropriate chunks of data where close to their destination place.
     // Finally squeeze data from respective dimensions.
     shared_ptr<Node> flat_node;
+    Shape dispersed_shape{n_dim};
+    for (int i = 0; i < spatial_dims; ++i)
+    {
+        dispersed_shape.push_back(bs);
+    }
+    const auto no_spatial_dims = data_shape.size() - spatial_dims;
```
Comment (tsocha) [inline src/ngraph/op/fused/depth_to_space.cpp]: I'm not a fan of this variable please remove/rename it.
This variable will be always equal to `2`.

File: src/ngraph/op/fused/space_to_depth.cpp [hunk 4]
```diff
@@ -43,57 +44,89 @@ op::SpaceToDepth::SpaceToDepth(const Output<Node>& data, const std::string& mode
...
+        data_shape.insert(data_shape.begin(), 1);
+        data = builder::reshape(data, data_shape);
     }
 
-    NGRAPH_CHECK((h % m_blocksize == 0 && w % m_blocksize == 0 && m_blocksize > 0),
-                 "SpaceToDepth: The width and height axes size must be a multiple of ",
-                 "squared block_size attribute value");
+    const size_t n_dim = data_shape.at(0);
+    const size_t c_dim = data_shape.at(1);
+    const size_t no_spatial_dims = 2;
```
Comment (tsocha) [inline src/ngraph/op/fused/space_to_depth.cpp]: Could you rename it to spatial_dims_index or something like this?

File: src/ngraph/op/fused/space_to_depth.cpp [hunk 5]
```diff
@@ -43,57 +44,89 @@ op::SpaceToDepth::SpaceToDepth(const Output<Node>& data, const std::string& mode
...
+                     "The dimension on position: ",
+                     i,
+                     " equal to: ",
+                     data_shape.at(i),
+                     " must be a multiple of m_blocksize: ",
+                     m_blocksize);
+    }
 
-    // First we have to disperse the data from height and width channels, then
+    // First we have to disperse the data from spatial channels, then
```
Comment (tsocha) [inline src/ngraph/op/fused/space_to_depth.cpp]:
```suggestion
    // First we have to disperse the data from spatial dimensions, then
```

File: src/ngraph/op/fused/space_to_depth.cpp [hunk 6]
```diff
@@ -43,57 +44,89 @@ op::SpaceToDepth::SpaceToDepth(const Output<Node>& data, const std::string& mode
...
+    auto flat_node = builder::reshape(data, dispersed_shape);
+    vector<size_t> axes_order{0};
+    for (size_t i = 0, j = 3; i < spatial_dims; ++i, j += 2)
+    {
+        axes_order.push_back(j);
+    }
+    for (size_t i = 0, j = 2; i < spatial_dims; ++i, j += 2)
+    {
+        axes_order.push_back(j);
+    }
```
Comment (tsocha) [inline src/ngraph/op/fused/space_to_depth.cpp]: An extra comment would be helpful.

File: src/ngraph/op/fused/depth_to_space.cpp (hunk 1 above)
Comment (mbencer) [reply src/ngraph/op/fused/depth_to_space.cpp]: Done

File: src/ngraph/op/fused/depth_to_space.cpp (hunk 2 above)
Comment (mbencer) [reply src/ngraph/op/fused/depth_to_space.cpp]: Done

File: src/ngraph/op/fused/space_to_depth.cpp (hunk 4 above)
Comment (mbencer) [reply src/ngraph/op/fused/space_to_depth.cpp]: I use `spatial_dim_index ` both for `DepthToSpace` and `SpaceToDepth`

File: src/ngraph/op/fused/space_to_depth.cpp (hunk 5 above)
Comment (mbencer) [reply src/ngraph/op/fused/space_to_depth.cpp]: Done

File: src/ngraph/op/fused/depth_to_space.cpp (hunk 3 above)
Comment (mbencer) [reply src/ngraph/op/fused/depth_to_space.cpp]: I used `spatial_dim_index`

File: src/ngraph/op/fused/space_to_depth.cpp (hunk 6 above)
Comment (mbencer) [reply src/ngraph/op/fused/space_to_depth.cpp]: Done

File: src/ngraph/op/fused/depth_to_space.cpp [hunk 7]
```diff
@@ -46,59 +47,99 @@ op::DepthToSpace::DepthToSpace(const Output<Node>& data,
...
-    NGRAPH_CHECK((c % (m_blocksize * m_blocksize) == 0 && m_blocksize > 0),
-                 "SpaceToDepth: The depth axis size must be a multiple of ",
-                 "squared block_size attribute value.");
+    const size_t n_dim = data_shape.at(0);
+    const size_t c_dim = data_shape.at(1);
+    const size_t spatial_dim_index = 2;
+    const size_t spatial_dims = data_shape.size() - spatial_dim_index;
+    const auto c_dim_divider = static_cast<int>(std::pow(m_blocksize, spatial_dims));
+
+    NGRAPH_CHECK(m_blocksize > 0 && c_dim % c_dim_divider == 0,
```
Comment (diyessi) [inline src/ngraph/op/fused/depth_to_space.cpp]: NODE_CHECK will give a better error message.

File: src/ngraph/op/fused/space_to_depth.cpp [hunk 8]
```diff
@@ -43,57 +44,91 @@ op::SpaceToDepth::SpaceToDepth(const Output<Node>& data, const std::string& mode
 NodeVector op::SpaceToDepth::decompose_op() const
 {
     auto data = input_value(0);
-    const Shape& data_shape = data.get_shape();
+    auto data_shape = data.get_shape();
 
-    // Set default values to each dimension to be able to work with both 3D or 4D data.
-    size_t n{1}, c{1}, h{1}, w{1};
+    NGRAPH_CHECK((data_shape.size() >= 3),
```
Comment (diyessi) [inline src/ngraph/op/fused/space_to_depth.cpp]: NODE_CHECK for these.

File: src/ngraph/op/fused/depth_to_space.cpp (hunk 7 above)
Comment (mbencer) [reply src/ngraph/op/fused/depth_to_space.cpp]: Done

File: src/ngraph/op/fused/space_to_depth.cpp (hunk 8 above)
Comment (mbencer) [reply src/ngraph/op/fused/space_to_depth.cpp]: Done

File: src/ngraph/op/fused/depth_to_space.cpp [hunk 9]
```diff
@@ -46,59 +47,101 @@ op::DepthToSpace::DepthToSpace(const Output<Node>& data,
...
+    // ..., DK])
+    // x'' = transpose(x', [0,  1,  K + 2, 2, K + 3, 3, K + 4, 4, ..., K + (K + 1), K + 1])
+    // y = reshape(x'', [N, C / (block_size ^ K), D1 * block_size, D2 * block_size, D3 * block_size,
+    // ..., DK * block_size])
     case DepthToSpaceMode::DEPTH_FIRST:
     {
-        flat_node = builder::reshape(data, Shape{n, c_flat, bs, bs, h, w});
-        flat_node = builder::reorder_axes(flat_node, {0, 1, 4, 2, 5, 3});
+        dispersed_shape.insert(dispersed_shape.begin() + 1, c_flat);
+        flat_node = builder::reshape(data, dispersed_shape);
```
Comment (GlebKazantaev) [inline src/ngraph/op/fused/depth_to_space.cpp]: Will it produce V1::Reshape? 
Due to we have to decompose this op in conversion layer the best way is to reuse operation decomposition method for both DepthToSpace and SpaceToDepth.

Comment (mbencer) [reply src/ngraph/op/fused/depth_to_space.cpp]: As agreed, I'll add producing `Reshape:v1` in separate PR

**Pull request merged**